libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/DecodeUtils.kt

```diff
@@ -0,0 +1,55 @@
+@file:JvmName("DecodeUtils")
+
+package com.mapbox.navigation.base.utils
+
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.core.constants.Constants
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.navigation.base.internal.utils.CacheResultUtils.cacheResult
+
+private val polylineDecodeCache = PolylineUtils::decode.cacheResult(maxSize = 50)
```

Comment:
This is too many to hold in the cache. I can't think of any reason to hold more than 3 routes, one primary and up to two alternatives.  Any more than this is using up too much memory with no benefit IMO.

Reply:
This cache is also used for steps, which have their own geometries, and there can be much more than 3 steps. 
I will try to split caches somehow, or store entries based on their length.

Reply:
The route line performance is hyper-dependent on the route being cached. The DirectionsRoute should have its own cache with a size of 3 to strike a balance between performance and memory consumption.  Having a large generalized cache will lead to storing routes unnecessarily and perhaps the primary route being removed from the cache if it's overused.  I suggest creating one or more caches based on stronger types than strings like a DirectionsRoute cache and perhaps a LegStep cache.

Reply:
I think we should measure that - having `DirectionsRoute` as a key in the cache still makes us do the equality check on it, which means each underlying step, the same as the current proposal does. The only overhead we're introducing is the iteration through the list. Let's see whether this actually impacts performance or not.

If we need to, we can maybe reserve some space for the route lines that we know are in use, so that developers do not overwrite that portion of the cache.

Reply:
One of my concerns is storing long routes unnecessarily. Take an extreme case where somebody initially gets 3 routes to choose from across the U.S. then searches for a different destination near their first one across the country. That will generate (potentially) 3 new routes which in this case would get cached so now that's 6 in memory and so on.  An edge case but not entirely unrealistic.